dispatcher/hub/commonHub.go

```diff
@@ -73,45 +63,30 @@ func checkArgs(args ArgsCommonHub) error {
 	return nil
 }
 
-// Run is launched as a goroutine and listens for events on the exposed channels
-// TODO: use protection from triggering multiple times
-func (ch *commonHub) Run() error {
+// RegisterListener creates a goroutine and listens for WS register events
+func (ch *commonHub) RegisterListener() error {
+	ch.mutState.Lock()
+	defer ch.mutState.Unlock()
+
+	if ch.cancelFunc != nil {
+		return common.ErrLoopAlreadyStarted
+	}
+
 	var ctx context.Context
 	ctx, ch.cancelFunc = context.WithCancel(context.Background())
 
-	go ch.run(ctx)
+	go ch.registerListener(ctx)
 
 	return nil
 }
 
-func (ch *commonHub) run(ctx context.Context) {
+func (ch *commonHub) registerListener(ctx context.Context) {
 	for {
 		select {
 		case <-ctx.Done():
-			log.Debug("commonHub is stopping...")
 			return
-
-		case events := <-ch.broadcast:
-			ch.handleBroadcast(events)
-
-		case revertEvent := <-ch.broadcastRevert:
-			ch.handleRevertBroadcast(revertEvent)
-
-		case finalizedEvent := <-ch.broadcastFinalized:
-			ch.handleFinalizedBroadcast(finalizedEvent)
-
-		case txsEvent := <-ch.broadcastTxs:
-			ch.handleTxsBroadcast(txsEvent)
-
-		case txsEvent := <-ch.broadcastBlockEventsWithOrder:
-			ch.handleBlockEventsWithOrderBroadcast(txsEvent)
-
-		case scrsEvent := <-ch.broadcastScrs:
-			ch.handleScrsBroadcast(scrsEvent)
-
 		case dispatcherClient := <-ch.register:
 			ch.registerDispatcher(dispatcherClient)
```

Comment:
why bother using a go-routine ran process loop just to handle the register/unregister event and we also use mutex operations on the ch.registerDispatcher and ch.unregisterDispatcher functions? The go routines & the 2 channels can be safely removed. The function RegisterEvent can call ch.registerDispatcher  and UnregisterEvent can call directly ch.unregisterDispatcher

Reply:
right, removed the loop